rosbag2_transport/src/rosbag2_transport/bag_rewrite.cpp

```diff
@@ -0,0 +1,199 @@
+  std::map<std::string, std::vector<std::string>> input_topics;
+  std::unordered_map<std::string, YAML::Node> input_topics_qos_profiles;
+  std::unordered_map<std::string, std::string> input_topics_serialization_format;
+
+  for (const auto & input_bag : input_bags) {
+    auto bag_topics_and_types = input_bag->get_all_topics_and_types();
+    for (const auto & topic_metadata : bag_topics_and_types) {
+      const std::string & topic_name = topic_metadata.name;
+      input_topics.try_emplace(topic_name);
+      input_topics[topic_name].push_back(topic_metadata.type);
+      input_topics_serialization_format[topic_name] = topic_metadata.serialization_format;
+
+      // Gather all offered qos profiles from all inputs
+      input_topics_qos_profiles.try_emplace(topic_name);
+      YAML::Node & all_offered = input_topics_qos_profiles[topic_name];
+      YAML::Node offered_qos_profiles = YAML::Load(topic_metadata.offered_qos_profiles);
+      for (auto qos : offered_qos_profiles) {
+        all_offered.push_back(qos);
+      }
+    }
+  }
+
+  for (const auto & [writer, record_options] : output_bags) {
+    rosbag2_transport::TopicFilter topic_filter{record_options};
+    auto filtered_topics_and_types = topic_filter.filter_topics(input_topics);
+
+    // Done filtering - set up writer
+    for (const auto & [topic_name, topic_type] : filtered_topics_and_types) {
+      rosbag2_storage::TopicMetadata topic_metadata;
+      topic_metadata.name = topic_name;
+      topic_metadata.type = topic_type;
+
+      // Take source serialization format for the topic if output format is unspecified
+      if (record_options.rmw_serialization_format.empty()) {
+        topic_metadata.serialization_format = input_topics_serialization_format[topic_name];
+      } else {
+        topic_metadata.serialization_format = record_options.rmw_serialization_format;
+      }
+
+      std::stringstream qos_profiles;
+      qos_profiles << input_topics_qos_profiles[topic_name];
+      topic_metadata.offered_qos_profiles = qos_profiles.str();
+      writer->create_topic(topic_metadata);
+
+      filtered_outputs.try_emplace(topic_name);
+      filtered_outputs[topic_name].push_back(writer.get());
+    }
+  }
+
+  return filtered_outputs;
+}
+
+void perform_rewrite(
+  const std::vector<std::unique_ptr<rosbag2_cpp::Reader>> & input_bags,
+  const std::vector<
+    std::pair<std::unique_ptr<rosbag2_cpp::Writer>, rosbag2_transport::RecordOptions>
+  > & output_bags
+)
+{
+  if (input_bags.empty() || output_bags.empty()) {
+    throw std::runtime_error("Must provide at least one input and one output bag to rewrite.");
+  }
+
+  auto topic_outputs = setup_topic_filtering(input_bags, output_bags);
+
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> next_messages;
+  next_messages.resize(input_bags.size(), nullptr);
+
+  std::shared_ptr<rosbag2_storage::SerializedBagMessage> next_msg;
+  while (next_msg = get_next(input_bags, next_messages)) {
```

Comment:
This introduced a warning message into the buildfarm. See [nightly_osx_debug#2246](https://ci.ros2.org/view/nightly/job/nightly_osx_debug/2246/).

It's probably unintended usage, would you take a look? @emersonknapp

Reply:
It is intended usage - as noted above I wasn't 100% sure that OSX would be happy. Will add the parentheses suggested by Clang ASAP.

Reply:
#930